cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp

```diff
@@ -14,16 +14,101 @@
 * limitations under the License.
 *******************************************************************************/
 
+#define DAAL_SYCL_INTERFACE
+#define DAAL_SYCL_INTERFACE_USM
+#define DAAL_SYCL_INTERFACE_REVERSED_RANGE
+
+#include <src/algorithms/kmeans/oneapi/kmeans_dense_lloyd_batch_kernel_ucapi.h>
+
 #include "oneapi/dal/algo/kmeans/backend/gpu/train_kernel.hpp"
+#include "oneapi/dal/backend/interop/common_dpc.hpp"
+#include "oneapi/dal/backend/interop/error_converter.hpp"
+#include "oneapi/dal/backend/interop/table_conversion.hpp"
+#include "oneapi/dal/exceptions.hpp"
 
 namespace oneapi::dal::kmeans::backend {
 
+using std::int64_t;
+using dal::backend::context_gpu;
+
+namespace daal_kmeans = daal::algorithms::kmeans;
+namespace interop     = dal::backend::interop;
+
+template <typename Float>
+using daal_kmeans_lloyd_dense_ucapi_kernel_t =
+    daal_kmeans::internal::KMeansDenseLloydBatchKernelUCAPI<Float>;
+
 template <typename Float>
 struct train_kernel_gpu<Float, method::lloyd_dense> {
     train_result operator()(const dal::backend::context_gpu& ctx,
                             const descriptor_base& params,
                             const train_input& input) const {
-        return train_result();
+        if (!(input.get_initial_centroids().has_data())) {
+            throw domain_error("Input initial_centroids should not be empty");
+        }
+
+        auto& queue = ctx.get_queue();
+        interop::execution_context_guard guard(queue);
+
+        const auto data = input.get_data();
+
+        const int64_t row_count    = data.get_row_count();
+        const int64_t column_count = data.get_column_count();
+
+        const int64_t cluster_count       = params.get_cluster_count();
+        const int64_t max_iteration_count = params.get_max_iteration_count();
+        const double accuracy_threshold   = params.get_accuracy_threshold();
+
+        daal_kmeans::Parameter par(cluster_count, max_iteration_count);
+        par.accuracyThreshold = accuracy_threshold;
+
+        auto arr_data        = row_accessor<const Float>{ data }.pull(queue);
+        const auto daal_data = interop::convert_to_daal_sycl_homogen_table(queue,
+                                                                           arr_data,
+                                                                           data.get_row_count(),
+                                                                           data.get_column_count());
+
+        auto arr_initial_centroids =
+            row_accessor<const Float>{ input.get_initial_centroids() }.pull(queue);
+
+        array<Float> arr_centroids = array<Float>::empty(queue, cluster_count * column_count);
+        array<int> arr_labels      = array<int>::empty(queue, row_count);
+        array<Float> arr_objective_function_value = array<Float>::empty(queue, 1);
```

Comment:
In old interests, all objects of size 1 are only created on the host. Do we want to change the logic?

Reply:
It seems like performance improvement without objectives. I think, we can do it consistently with other memory allocations now (=allocate shared USM), but if these will lead to perf degradation, then fix it. Moreover, it is better to allocate such single-sized objects on a stack rather than on the heap.